benchmarks/persistence/persistence.py

```diff
@@ -0,0 +1,228 @@
+    """The name of the benchmark for which the path should be constructed."""
+
+    branch: str = field(validator=instance_of(str), init=False)
+    """The branch currently checked out."""
+
+    latest_baybe_tag: str = field(validator=instance_of(str))
+    """The latest BayBE version tag."""
+
+    execution_date_time: datetime = field(validator=instance_of(datetime))
+    """The date and time when the benchmark was executed."""
+
+    commit_hash: str = field(validator=instance_of(str))
+    """The latest commit hash of the current checkout branch."""
+
+    @branch.default
+    def _default_branch(self) -> str:
+        repo = Repo(search_parent_directories=True)
+        current_branch = repo.active_branch.name
+        return current_branch
+
+    def _string_sanitizer(self, string: str) -> str:
+        """Replace not allowed chars for filesystems in a string."""
+        ALLOWED_CHARS = (
+            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+        )
+        return "".join([char if char in ALLOWED_CHARS else "-" for char in string])
+
+    def get_path(self, strategy: PathStrategy) -> str:
+        """Construct the path of a result object.
+
+        Creates the path depending on the chosen strategy by concatenating
+        the experiment identifier, the branch, the BayBE-version,
+        the start date without time and the commit hash.
+
+        Args:
+            strategy: The strategy to construct the path.
+
+        Returns:
+            The path to persist the object. Can be a file name or
+            a folder path as a string.
+        """
+        separator = "/" if strategy == PathStrategy.HIERARCHICAL else "_"
+
+        file_usable_date = self.execution_date_time.strftime("%Y-%m-%d")
+        components = [
+            self.benchmark_name,
+            self.branch,
+            self.latest_baybe_tag,
+            file_usable_date,
+            self.commit_hash,
+        ]
+
+        sanitized_components = [
+            self._string_sanitizer(component) for component in components
+        ]
+        path = separator.join(sanitized_components) + separator + "result.json"
+
+        return path
+
+
+class PersistenceHandlingInterface(Protocol):
+    """Interface for interacting with storage."""
+
+    def write_object(self, path_constructor: PathConstructor, object: dict) -> None:
+        """Store a JSON serializable dict according to the path.
+
+        If the respective object exists, it will be overwritten.
+
+        Args:
+            path_constructor: The PathConstructor to create the path for the object.
+            object: The object to be persisted.
+        """
+
+
+@define
+class S3PersistenceHandler:
+    """Class for persisting objects in an S3 bucket."""
+
+    _bucket_name: str = field(validator=instance_of(str), init=False)
+    """The name of the S3 bucket where the results are stored."""
+
+    _object_session: Session = boto3.session.Session()
```

Comment:
I guess this should use a factory, no?
```suggestion
    _object_session: Session = field(factory=boto3.session.Session)
```